### PR TITLE
 Replace Vec<u8> with [u8; 32] for contract storage key

### PR DESCRIPTION
**Pull request by ascjones**
Contract storage keys are 32 bytes.

Introduces an additional function `read_sandbox_memory_into_buf` which avoids allocating a vec when we know the size of the data up front.

### CODE REVIEW

**Comment by gavofyork** [comment]
is there an issue for this?

**Comment by ascjones** [comment]
Don't think so. It was a suggestion from @pepyakin.

File: srml/contract/src/wasm/runtime.rs [hunk 1]
```diff
@@ -177,6 +177,26 @@ fn read_sandbox_memory<E: Ext>(
...
+fn read_sandbox_memory_into_buf<E: Ext>(
+	ctx: &mut Runtime<E>,
+	ptr: u32,
+	buf: &mut [u8],
+) -> Result<(), sandbox::HostError> {
+	charge_gas(ctx.gas_meter, ctx.schedule, RuntimeToken::ReadMemory(buf.len() as u32))?;
+
+	ctx.memory().get(ptr, buf)?;
+
+	Ok(())
```
Comment (bkchr) [inline srml/contract/src/wasm/runtime.rs]: I think we could remove this and the `?;` above?

File: srml/contract/src/wasm/runtime.rs [hunk 2]
```diff
@@ -228,14 +248,15 @@ define_env!(Env, <E: Ext>,
 	//   where the value to set is placed. If `value_non_null` is set to 0, then this parameter is ignored.
 	// - value_len: the length of the value. If `value_non_null` is set to 0, then this parameter is ignored.
 	ext_set_storage(ctx, key_ptr: u32, value_non_null: u32, value_ptr: u32, value_len: u32) => {
-		let key = read_sandbox_memory(ctx, key_ptr, 32)?;
+		let mut key = [0; 32];
```
Comment (bkchr) [inline srml/contract/src/wasm/runtime.rs]: I would at least make it `key: StorageKey` to make sure we are informed when the length of the key changes.

File: srml/contract/src/wasm/runtime.rs [hunk 3]
```diff
@@ -247,7 +268,8 @@ define_env!(Env, <E: Ext>,
 	// - key_ptr: pointer into the linear memory where the key
 	//   of the requested value is placed.
 	ext_get_storage(ctx, key_ptr: u32) -> u32 => {
-		let key = read_sandbox_memory(ctx, key_ptr, 32)?;
+		let mut key = [0; 32];
```
Comment (bkchr) [inline srml/contract/src/wasm/runtime.rs]: Same as above.

**Pull request merged**